src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs

```diff
@@ -595,37 +595,87 @@ internal static int CompareOrdinalIgnoreCase(ref char strA, int lengthA, ref cha
+                while ((uint)length >= 2)
                 {
-                    length--;
-                    charA = ref Unsafe.Add(ref charA, 1);
-                    charB = ref Unsafe.Add(ref charB, 1);
+                    // Read 2 chars (32 bits) at a time from each string
+                    uint valueA = Unsafe.ReadUnaligned<uint>(ref Unsafe.As<char, byte>(ref Unsafe.AddByteOffset(ref charA, byteOffset)));
+                    uint valueB = Unsafe.ReadUnaligned<uint>(ref Unsafe.As<char, byte>(ref Unsafe.AddByteOffset(ref charB, byteOffset)));
+
+                    if (!Utf16Utility.AllCharsInUInt32AreAscii(valueA | valueB))
+                    {
+                        goto NonAscii; // one of the inputs contains non-ASCII data
+                    }
+
+                    // Generally, the caller has likely performed a first-pass check that the input strings
+                    // are likely equal. Consider a dictionary which computes the hash code of its key before
+                    // performing a proper deep equality check of the string contents. We want to optimize for
+                    // the case where the equality check is likely to succeed, which means that we want to avoid
+                    // branching within this loop unless we're about to exit the loop, either due to failure or
+                    // due to us running out of input data.
+
+                    if (!Utf16Utility.UInt32OrdinalIgnoreCaseAscii(valueA, valueB))
+                    {
+                        return false;
+                    }
+
+                    byteOffset += 2;
+                    length -= 2;
                 }
-                else
+
+                if (length != 0)
                 {
-                    return false;
+                    Debug.Assert(length == 1);
+
+                    uint valueA = Unsafe.AddByteOffset(ref charA, byteOffset);
+                    uint valueB = Unsafe.AddByteOffset(ref charB, byteOffset);
+
+                    if ((valueA | valueB) > 0x7Fu)
+                    {
+                        goto NonAscii; // one of the inputs contains non-ASCII data
+                    }
+
+                    if (valueA == valueB)
+                    {
+                        return true;
+                    }
+
+                    valueA |= 0x20u;
+                    return ((uint)(valueA - 'a') <= (uint)('z' - 'a'))
+                        && (valueA == (valueB | 0x20u));
                 }
-            }
 
-            if (length == 0)
+                Debug.Assert(length == 0);
                 return true;
 
-            Debug.Assert(!GlobalizationMode.Invariant);
+            NonAscii:
+                return CompareStringOrdinalIgnoreCase(ref Unsafe.AddByteOffset(ref charA, byteOffset), length, ref Unsafe.AddByteOffset(ref charB, byteOffset), length) == 0;
+            }
+            else
+            {
+                while (length != 0)
```

Comment:
The same code above would work even in this GlobalizationMode.Invariant case, except the ASCII checks would fail? I guess if GlobalizationMode.Invariant became important/common we could optimize it similarly.